File: interpreters/pigs.js
let variables = {0: 0};
let waiting = false;
let inp = undefined;
let commands, c, pv;
let interval;

function run() {
    let i, j, J;
    let b = false;
    document.getElementById("status").innerHTML = "Running...";
    if (commands)
    interval = setInterval(() => {
        if (c < commands.length) {
            i = commands[c][0];
            j = commands[c].substring(1, commands[c].length);
            j = parseInt(j, 2);
            initvar(j);
            J = variables[j];
            switch (i) {
                case "P":
                    try {
                        output(String.fromCodePoint(J));
                    } catch {}
                    break;
                case "I":
                    if (inp === undefined) {
                        if (!waiting) {
                            output('<textarea class="input" rows="1" cols="5"></textarea><button id="submit" onclick="sub();" onchange="">Submit</button>');
                            waiting = true;
                        }
                    } else {
                        for (var k = 0; k < inp.length; k++) {
                            idx = J + k;
                            clamp(idx);
                            variables[idx] = inp.codePointAt(k);
                        }
                        inp = undefined;
                    }
                    break;
                case "G":
                    c = J - 1;
                    break;
                case "S":
                    pv = J;
                    initvar(pv);
                    break;
                case "+":
                    variables[pv] += J;
                    clamp(variables[pv]);
                    break;
                case "-":
                    variables[pv] -= J;
                    clamp(variables[pv]);
                    break;
                case "=":
                    variables[pv] = Number(variables[pv] === J);
                    break;
                case ":":
                    initvar(J);
                    variables[pv] = variables[J];
                    break;
                case "#":
                    variables[pv] = j;
                    clamp(variables[pv]);
                    break;
            }
            if (!waiting) {
                c++;
            } else {
                b = true;
                document.getElementById("status").innerHTML = "Waiting for input...";
            }
        } else {
            b = true;
            document.getElementById("status").innerHTML = "";
        }
        if (b) {clearInterval(interval); b = false;}
    }, 0);
}

function sub() {
    let inputs = document.getElementsByClassName("input");
    let input = inputs[inputs.length - 1];
    input.readOnly = true;
    input.classList.add("inputted")
    input.innerHTML = input.value;
    inp = input.value;
    waiting = false;
    document.getElementById("submit").remove()
    run();
}

function reset() {
    document.getElementById("output").innerHTML = "";
    variables = {0: 0};
    commands = document.getElementById("program").value.replace(/[^PIGS+-=:#01]/g, "").match(/([PIGS+-=:#])([01]+)/g);
    c = 0;
    pv = 0;
}

function output(x) {
    out = document.getElementById("output");
    document.getElementById("output").innerHTML += x;
    out.scrollTop = out.scrollHeight;
}

let initvar = v => {if (!Object.keys(variables).includes(v.toString())) variables[v] = 0;}
let clamp = n => {n = n % (2 ** 32);}

File: interpreters/pigs.py
import re

def initvar(v):
    if v not in variables:
        variables[v] = 0

def clamp(n): n = n % (2 ** 32)


program = input()
program = re.sub('[^PIGS+-=:#01]', '', program)
#print('Compressed Program:', program + '\n')

variables = {0: 0}
commands = [i for i in re.findall('([PIGS+-=:#])([01]+)', program) if i[0] in 'PIGS+-=:#']
#print(commands)

c = 0
pv = 0

while c < len(commands):
    i, j = commands[c]
    j = int(j, 2)
    initvar(j)
    J = variables[j]
    if i == 'P':
        try:
            print(chr(J), end='', flush=True)
        except:
            pass
    elif i == 'I':
        inp = input()
        for k in range(len(inp)):
            idx = J + k
            clamp(idx)
            variables[idx] = ord(inp[k])
    elif i == 'G':
        c = J - 1
    elif i == 'S':
        pv = J
        initvar(pv)
    elif i == '+':
        variables[pv] += J
        clamp(variables[pv])
    elif i == '-':
        variables[pv] -= J
        clamp(variables[pv])
    elif i == '=':
        variables[pv] = int(variables[pv] == J)
    elif i == ':':
        initvar(J)
        variables[pv] = variables[J]
    elif i == '#':
        variables[pv] = j
        clamp(variables[pv])
    c += 1
